### PR TITLE
[WebAssembly] Fix shared library loading

### PR DESCRIPTION
**Pull request by sbc100**
Loading of shared libraries was broken back in:
https://github.com/jfbastien/musl/commit/39fde6042ee22bc75556f4ba063775b3e457e471

This change restores the ability for shared libraries to
be loaded such that they can share memory and provide
function exports to the main executable.

However the executable that libraries still need to built
 in a specific way (i.e. all imported the same sized memory)
and their static globals need to be in the different address
range.    This has the complication that the executables won't
run the wasm spec interpreter (since it doesn't pass in a
memory to the executables it runs).   So although I'm fixing this
here in the musl repo I'm also proposiing that we remove the
continuous testing of this configurations on the wasm waterfall
in favor of concentrating on the statically linking lld case:
https://github.com/WebAssembly/waterfall/pull/271

### CODE REVIEW

**Comment by jfbastien** [comment]
I'm not sure I understand: why don't we share the memory? That used to be what this code did, before we actually standardized dynamic linking. In fact, this code was what motivated dynamic linking.

I agree the code is totally wrong now (I think you and I discussed this is a previous PR).

**Comment by sbc100** [comment]
Looks like shared memory stopped working with this commit:
https://github.com/jfbastien/musl/commit/39fde6042ee22bc75556f4ba063775b3e457e471

Prior to this the heap was shared between all loaded modules.. after this change each module exports its own heap and the JS `heap` variable is set to the last loaded one..

**Comment by sbc100** [comment]
I opened https://github.com/jfbastien/musl/issues/19 to try to address the shared memory issue.

**Comment by jfbastien** [comment]
Sounds good. Should we not merge this PR then?

**Comment by sbc100** [comment]
Well if we merge this PR then symbol linking starts working again, but without shared memory, so we get a whole host of different failures on the waterfall.. along with some new successes.. so I think no, we should fix both symbol linking and memory linking in the same PR.  I'll see what I can do to improve this one.

**Comment by sbc100** [comment]
I think this should be ready to merge now... ptal

File: arch/wasm32/wasm.js [hunk 1]
```diff
@@ -959,9 +964,14 @@ if (arguments[0] == '--dump-ffi-symbols') {
       ? new Uint8Array(readbuffer(file_path))
       : read(file_path, 'binary');
     instance = new WebAssembly.Instance(new WebAssembly.Module(buf), ffi)
-    heap = instance.exports.memory.buffer;
-    heap_uint8 = new Uint8Array(heap);
-    heap_size_bytes = heap.byteLength;
+    // For the application exports its memory, we use that rather than
+    // the one we created. In this way this code works with modules that
+    // import or export thier memory.
```
Comment (jfbastien) [inline arch/wasm32/wasm.js]: Typo "their"

**Review by jfbastien — APPROVED** [review]

**Pull request merged**